python/lance/_lib.pyx

```diff
@@ -138,6 +146,56 @@ cdef class LanceFileFormat(FileFormat):
         return LanceFileWriteOptions.wrap(self.format.DefaultWriteOptions())
 
 
+cdef extern from "lance/arrow/updater.h" namespace "lance::arrow" nogil:
+    cdef cppclass CUpdater "::lance::arrow::Updater":
+        CResult[shared_ptr[CRecordBatch]] Next();
+
+        CStatus UpdateBatch(const shared_ptr[CArray] arr);
+
+        CResult[shared_ptr[CLanceDataset]] Finish();
+
+    cdef cppclass CUpdaterBuilder "::lance::arrow::UpdaterBuilder":
+        CResult[shared_ptr[CUpdater]] Finish();
+
+
+cdef class Updater:
+    cdef shared_ptr[CUpdater] sp_updater
+
+    @staticmethod
+    cdef wrap(const shared_ptr[CUpdater]& up):
+        cdef Updater self = Updater.__new__(Updater)
+        self.sp_updater = move(up)
+        return self
+
+    def next(self) -> Optional[pyarrow.Table]:
+        cdef shared_ptr[CRecordBatch] c_batch
+        c_batch = GetResultValue(self.sp_updater.get().Next())
+        if c_batch.get() == NULL:
+            return None
+        batch = pyarrow_wrap_batch(c_batch)
+        return pyarrow.Table.from_batches([batch])
+
+    def update_batch(self, data: pyarrow.Array):
+        cdef shared_ptr[CArray] arr = pyarrow_unwrap_array(data)
+        with nogil:
+            check_status(self.sp_updater.get().UpdateBatch(arr))
+
+    def finish(self):
+        cdef shared_ptr[CLanceDataset] new_dataset = GetResultValue(
+            self.sp_updater.get().Finish()
+        )
+        return FileSystemDataset.wrap(static_pointer_cast[CDataset, CLanceDataset](new_dataset))
+
+    def __iter__(self):
```

Comment:
Why do you need both __iter__ and next?

Reply:
Do you mean `next()` or `__next__()` here?